[FIX] odoo/fields.py: fix removing translations on writing ''

yelizariev opened the pull request:
cache_value could be empty string rather than None. I'm not sure why it was
different before.

STEPS:
* install an app that adds menu Products (e.g. Sales)
* activate second language and switch to it
* create new product, set some value to field description ("Internal Notes"),
save
* click edit, remove description, save

BEFORE: the value is not removed

AFTER: value is empty, translations are removed

---

https://github.com/odoo/odoo/commit/543a5523a30c993f60eba3c6d56d54dea2857eb3#
opw-2426724

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:




--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/64241).

yelizariev commented:
@mart-e @rco-odoo

mart-e commented:
Why wasn't you scenario detected by this test? Maybe it needs to be updated

https://github.com/odoo/odoo/blob/99531f5dd8966fc4ad89d1e9c2886ad701527f3a/odoo/addons/base/tests/test_translate.py#L505

yelizariev commented:
it writes false and not empty string

https://github.com/odoo/odoo/blob/99531f5dd8966fc4ad89d1e9c2886ad701527f3a/odoo/addons/base/tests/test_translate.py#L536

yelizariev commented:
@mart-e 
I've updated the test to write empty string as it's done in backend

![image](https://user-images.githubusercontent.com/186131/104167414-8ef1b000-53fc-11eb-91ad-ae50d1ae82ce.png)

Also, I removed creating translation and just write record with fr_FR context

yelizariev commented:
Though, removing translation makes test name obsolete. I'll revert this part

MiquelRForgeFlow commented on odoo/fields.py:
```diff
@@ -1448,7 +1448,7 @@ def write(self, records, value):
                 # cache_value is already in database format
                 towrite[rid][self.name] = cache_value
-            if self.translate is True and cache_value is not None:
+            if self.translate is True and not cache_value:
```
```suggestion
            if self.translate is True and cache_value:
```

right?

yelizariev replied:
Thanks. I was about to check why other tests don't work :smiley:

mart-e commented:
>  I've updated the test to write empty string as it's done in backend

Hmm interresting, actually emptying the value on a `fields.Char` writes `False` (as covered by the test) but on `fields.Text`, the web client writes `''`.
Not sure the difference is intended? (cc @aab-odoo maybe ?)

For the test,  we already have `.with_context(lang='fr_FR').write({...: False})`, maybe we should keep it and add another case `.with_context(lang='en_US').write({...: ""})` to cover all possibilities?

aab-odoo commented:
> Hmm interresting, actually emptying the value on a `fields.Char` writes `False` (as covered by the test) but on `fields.Text`, the web client writes `''`.
> Not sure the difference is intended? (cc @aab-odoo maybe ?)

Where do you see that the backend behaves differently for char and text fields? It is probably not intended, or if it is, I can't remember why. Is it a problem?

yelizariev commented:
@aab-odoo it does behaves differently for chat and text fields (pass False and "" correspondingly)
![image](https://user-images.githubusercontent.com/186131/104451571-090f6980-55a2-11eb-9fe2-7e7505e57fe1.png)

yelizariev commented:
Problem or not, it's question for python framework team, not for Martin ;-)

yelizariev commented:
From my side, it's a problem until we merge  this PR

mart-e reviewed (approved):
Looks good for me :+1:  Is the second commit linked to the same issue?
@robodoo delegate+

yelizariev commented:
Second commit was pushed by mistake. Removed now

rco-odoo reviewed (approved):
Looks reasonable, but that's a semantic change, not a bug fix.

yelizariev commented:
@robodoo r+

sebalix commented:
@yelizariev @mart-e @rco-odoo We spotted an issue with this PR, all translations are removed if we set an empty string only for one of them, we needed to revert this commit on our side to not trigger "not null" constraint on some translatable fields (such as attribute values).
Steps to reproduce the issue:

**Before this PR** (it empties only the targeted translation and doesn't update the others):
```python
>>> env.context.get("lang")
'en_US'
>>> p = env["product.template"].create({"name": "EN translation"})
>>> p.name
'EN translation'
>>> p.with_context(lang="fr_CH").name = "FR translation"
>>> p.with_context(lang="it_IT").name = "IT translation"
>>> p.with_context(lang="de_DE").name = "DE translation"
>>> p.with_context(lang="fr_CH").name
'FR translation'
>>> p.with_context(lang="it_IT").name
'IT translation'
>>> p.with_context(lang="de_DE").name
'DE translation'
>>> p.with_context(lang="fr_CH").name = ""
>>> p.name
'EN translation'
>>> p.with_context(lang="fr_CH").name
''
>>> p.with_context(lang="it_IT").name
'IT translation'
>>> p.with_context(lang="de_DE").name
'DE translation'
```

**With this PR** (all translations are removed while setting an empty string for french translation only):
```python
>>> env.context.get("lang")
'en_US'
>>> p = env["product.template"].create({"name": "EN translation"})
>>> p.name
'EN translation'
>>> p.with_context(lang="fr_CH").name = "FR translation"
>>> p.with_context(lang="it_IT").name = "IT translation"
>>> p.with_context(lang="de_DE").name = "DE translation"
>>> p.with_context(lang="fr_CH").name
'FR translation'
>>> p.with_context(lang="it_IT").name
'IT translation'
>>> p.with_context(lang="de_DE").name
'DE translation'
>>> p.with_context(lang="fr_CH").name = ""
2021-03-03 12:24:50,815 1 INFO db-20210301-120058 odoo.models.unlink: User #1 deleted ir.translation records with IDs: [2674727, 2674728, 2674729]
>>> p.name
''
>>> p.with_context(lang="it_IT").name
''
>>> p.with_context(lang="de_DE").name
''
```
Can you have a look please?

yelizariev commented:
@sebalix
I'll take care on it. 
I think we need to add language filter for unlink

https://github.com/odoo/odoo/blob/24c3260714083934a02ae349954136ec8839d1c8/odoo/fields.py#L1508-L1512

yelizariev commented:
@sebalix 
actually, it's intended behavior. It's exactly was done by @mart-e in https://github.com/odoo/odoo/commit/543a5523a30c993f60eba3c6d56d54dea2857eb3

simahawk commented:
@yelizariev @mart-e it sounds more wrong than before... I'd expect complete deletion only if the language was the main one (eg: the source of the translation). But if I want to wipe a translation for a good reason (eg: the string in my language is the same as the one in English) why on heart shall you wipe them all? :thinking:

mart-e commented:
@simahawk if you want to explicitly remove one translation, you have to act on the `ir.translation` record directly, not on the referenced record. Remember there is no such thing as "main language".
In the interface, this means going to the popup list of translation and setting `""` as the value in the languages you want to discard.
If you write `""` directly in a text field, you should never get another value back after saving, this too confusing.
